src/Controls/src/SourceGen/RouteSourceGenerator.cs

```diff
@@ -0,0 +1,252 @@
+﻿using System.Linq;
+using System.Text;
+
+using Microsoft.CodeAnalysis;
+
+using static Microsoft.Maui.Controls.SourceGen.Helper;
+
+namespace Microsoft.Maui.Controls.SourceGen
+{
+	[Generator(LanguageNames.CSharp)]
+	public class RouteSourceGenerator : ISourceGenerator
+	{
+		static readonly string Indent = new string('\x20', 8);
+		const string Add = nameof(Add);
+		const string TryAdd = nameof(TryAdd);
+
+		const string AutoGeneratedHeaderText = """
+			//------------------------------------------------------------------------------
+			// <auto-generated>
+			//     This code was generated by a .NET MAUI source generator.
+			//
+			//     Changes to this file may cause incorrect behavior and will be lost if
+			//     the code is regenerated.
+			// </auto-generated>
+			//------------------------------------------------------------------------------
+			""";
+
+		public void Initialize(GeneratorInitializationContext context)
+		{
+			context.RegisterForSyntaxNotifications(() => new RouteSyntaxReceiver());
+		}
+
+		public void Execute(GeneratorExecutionContext context)
+		{
+			if (context.SyntaxReceiver is not RouteSyntaxReceiver syntaxReceiver)
+			{
+				return;
+			}
+
+			var mauiApp = syntaxReceiver.MauiApp;
+			var mauiStartup = syntaxReceiver.MauiStartup;
+			var routedPages = syntaxReceiver.RoutedPages;
+			var shellPages = syntaxReceiver.ShellPages;
+
+			if (shellPages.Count() == 0 || routedPages.Count() == 0)
+			{
+				return;
+			}
+
+			var routes = new StringBuilder();
+			var services = new StringBuilder();
+
+			var addComment = true;
+
+			foreach (var routedPage in routedPages.Where(x => x.Lifetime == Singleton))
+			{
+				foreach (var route in routedPage.Routes)
+				{
+					if (string.IsNullOrEmpty(route))
+					{
+						routes.AppendLine($"{Indent}Routing.RegisterRoute(nameof(global::{GetNamespace(routedPage.Type)}.{routedPage.Type.Identifier}), typeof(global::{GetNamespace(routedPage.Type)}.{routedPage.Type.Identifier}));");
+					}
+					else
+					{
+						routes.AppendLine($"{Indent}Routing.RegisterRoute({route}, typeof(global::{GetNamespace(routedPage.Type)}.{routedPage.Type.Identifier}));");
+					}
+				}
+
+				if (addComment)
+				{
+					addComment = false;
+					services.AppendLine($"// Singleton Services");
+				}
+
+				if (routedPage.ImplicitViewModel)
+				{
+					services.AppendLine($"{Indent}builder.Services.AddSingleton<{routedPage.Type.Identifier.ValueText.Replace("Page", "ViewModel")}>();");
+				}
+				else if (!string.IsNullOrEmpty(routedPage.ViewModelType))
+				{
+					services.AppendLine($"{Indent}builder.Services.AddSingleton<{routedPage.ViewModelType}>();");
+				}
```

Comment:
No need for the else clause.

Reply:
The reason for the else clause is that if both `ViewModelType` and `ImplicitViewModel` properties are defined, `ViewModelType` takes precedence as it is an explicit value. In fact, I've to reverse the condition such that it is evaluated first.

Reply:
Ok, that makes me worry, not for the current state of the code, but the chance that someone will come after us without the knowledge, will see the else and remove it. Without the context, the else makes no sense. 

For example, a jr dev joins the team, opens the file and their IDE tells them that the else does nor make sense and trusts the IDE, then we have a bug.

Maybe the following idea can help us make the code more robust for people in the future eager to simplify code (AKA: *me in 2025*).

We could add a extension method for the string builder, it takes the pattern and a bool the will let us know if we have to append or not:

```csharp
if (routedPage.ImplicitViewModel)
{
	services.AppendLine($"{Indent}builder.Services.AddSingleton<{routedPage.Type.Identifier.ValueText.Replace("Page", "ViewModel")}>();");
}
else 
{
	services.AppendLine($"{Indent}builder.Services.AddSingleton<{routedPage.ViewModelType}>();", shouldAppend: !string.IsNullOrEmpty(routedPage.ViewModelType));
}

// somewhere in a static class
public static StringBuilder AppendLine (this StringBuilder self, string? value, bool shouldAppend = true)
{
	if (! shouldAppend)
		return self;
	return self.AppendLine (value);
}
```

I noticed that this pattern is only used in ifs with the ImplicitViewModel and the ViewModelType so we could even just make the ApplendLine a simple helper method in the class.

Reply:
Understand your concern. Will look for ways to make it maintainable.

In addition, making minor changes to include the fully qualified namespace for the `ViewModel` type as it's auto-generated and there would be no provision for the user to do any changes here. Kindly bear with me, will notify you once it's committed.